chore(no-loop): use BFS to find loops

marstamm opened the pull request:
This uses [breadth-first search](https://en.wikipedia.org/wiki/Breadth-first_search), [cf. detailed rationale](https://github.com/camunda/bpmnlint-plugin-camunda-compat/pull/165#issuecomment-2196882155). As a result it reduces the time needed for traversing the diagram from [SUPPORT-22488](https://jira.camunda.com/browse/SUPPORT-22488) from ~1 minute to 6ms. 

Related to https://github.com/camunda/camunda-modeler/issues/4401

![image](https://github.com/camunda/bpmnlint-plugin-camunda-compat/assets/21984219/5bb01858-fd59-4ba5-8453-62d6571601f1)

Try it out with 
```
npx @bpmn-io/sr camunda/linting -l camunda/bpmnlint-plugin-camunda-compat#make-loop-detection-more-perfomant
```

philippfromme commented on rules/camunda-cloud/no-loop.js:
```diff
@@ -38,16 +38,17 @@ module.exports = skipInNonExecutableProcess(function() {
     }
 
-    const error = getFlowElements(node)
+    // Create subgrapg of nodes that can be part of an infinite loop
```
Change to `subgraph`

marstamm replied:
Done in [2dd0b78](https://github.com/camunda/bpmnlint-plugin-camunda-compat/pull/165/commits/2dd0b782a8e4bdd7fa22c3140aaa84d8f867e3f7)

nikku reviewed (commented):
Please explain in a nutshell what you did and why.

Consider to add a semi-big test case and assert that no computational complexity explosion happens.

marstamm commented:
I only reworked the traversal code of the rule. We are treating the diagram as a directed graph. Starting with the complete diagram, we follow these steps:

1. Create a subgraph that only contains elements that can be part of an infinite loop (eg. filter out ServiceTasks, ...)
2. On this subgraph, perform a breadth-first search. Should we find an element that we already visited, we found a loop
3. As we already filtered for elements that do no computing, we report the error

Steps 1 and 2 are new/reworked by me, rest is taken from the original lint rule.

marstamm commented:
We assume the worst case of a fully connected diagram with `V=n` elements and `E=n^2` edges. The previous Algorithm was:
 - For **every** element that could be part of a loop (n)
   - Traverse **all possible paths** starting from this element. In the worst case, this could be O(n!), I think.

We did not do any caching, so the same paths would be calculated multiple times for each new starting element.

In the worst case, this yields `O(n!)`, which is quite horrible. BFS is `O(V+E)`, so `O(n^2)`, with much lower complexity in real-world cases.

philippfromme reviewed (changes requested):
We found an issue that needs fixing before we can merge.

philippfromme commented:
Remembering visited nodes and returning early when visiting them again break in a case where

1. The first branch marked a gateway as visited

![image](https://github.com/camunda/bpmnlint-plugin-camunda-compat/assets/7633572/ac0d22f6-eb81-43a7-b16f-8a8a7d671aa6)

2. The second branch returns early when visiting that gateway

![image](https://github.com/camunda/bpmnlint-plugin-camunda-compat/assets/7633572/40e2cd4a-b50d-49c1-8420-5e1d4156ee30)

marstamm commented:
The Issue we have is that the loop detection works, but we ignore certain loops because it requires a Task in it to be "valid". Therefore, I think we should do more pre-processing so our graph can be traversed and any loop found is actually a loop. 

This would look like this

Step 0: Initial Diagram
![image](https://github.com/camunda/bpmnlint-plugin-camunda-compat/assets/21984219/f22f7fd1-e22e-4ecc-b813-ffe66ab24385)

Step 1: Remove Blocking elements (filter, O(n))
![image](https://github.com/camunda/bpmnlint-plugin-camunda-compat/assets/21984219/ba82fb86-b994-432d-a8e8-f68617748440)

Step 2:
Treat allowed but non-required elements as annotated edges. (BFS, O(n^2))
![image](https://github.com/camunda/bpmnlint-plugin-camunda-compat/assets/21984219/749cfa8c-b86c-4fa3-84dd-9d905e403038)

In the resulting graph, any Loop will be a valid loop.

This increases the complexity as we need to iterate over the complete graph twice, but will not change the O-complexity O(2*n^2) = O(n^2)

nikku commented:
@marstamm In your algorithm proposal (https://github.com/camunda/bpmnlint-plugin-camunda-compat/pull/165#issuecomment-2197131973), how would you handle gateway-only situations that can be loops, too:

![image](https://github.com/camunda/bpmnlint-plugin-camunda-compat/assets/58601/8823b263-1e49-469f-86b9-98dcbe2fbf5f)

marstamm commented:
Those will be ignored. The resulting "reduced" graph will be empty, because no required Element exists. 

I synced with Philipp on Friday, we want to align with the Zeebe implementation. For the Zeebe implementation, it is required for the loop to include a Blank-, Manual- or CallActivity

marstamm commented:
I added the 2-step process described in https://github.com/camunda/bpmnlint-plugin-camunda-compat/pull/165#issuecomment-2197131973 and added additional test-cases that cover the previously broken scenarios.

We now use a custom Graph Structure to make traversal and annotation of paths easier:
https://github.com/camunda/bpmnlint-plugin-camunda-compat/blob/85bd00872d5b6bc128276f7d197443bacabe2a41/rules/camunda-cloud/no-loop.js#L69-L74


All reported Loops now start at a Task or CallActivity. I could not detect any noticeable performance decreases to the previous implementation:
![image](https://github.com/camunda/bpmnlint-plugin-camunda-compat/assets/21984219/638ae4dc-4d49-4f1f-b211-7f38a2d48322)

philippfromme reviewed (approved):
Good job! ✅

nikku commented:
@marstamm Let's get this released and integrated.

marstamm commented:
Already on it 🏃

nikku commented on rules/camunda-cloud/no-loop.js:
```diff
@@ -56,51 +66,112 @@ module.exports = skipInNonExecutableProcess(function() {
+      connectNodes(graph, node);
+    }
+
+    return [ ...outgoing.keys() ].map(key => graph.get(key));
```
```suggestion
    return [ ...outgoing.keys() ].map(key => graph.get(key));
```

We could use `Array.from(...)` to not get into infinite loop issues here.

marstamm replied:
What do you mean by "infinite loop"? We could use `Array.from(iterable, mapFn)`, but I do not see where the current implementation creates a loop

marstamm replied:
If you are referring to the `Maximum call stack size exceeded` issue we had in the variable resolver, this was because `.push(...array)` and other functions will create a stack entry for every element. 
Creating new arrays using the spread operator is safe, as it is not a function call

nikku replied:
Yea, I referred to that instance of a problem. If we have a built-in utility, we could use it. That is what I'm saying, between the lines.

marstamm replied:
I created a follow-up here: https://github.com/camunda/bpmnlint-plugin-camunda-compat/pull/167

As the current implementation is not vulnerable to the Max Call stack issue, I see this as a simple code-style improvements which we do not need to release right away.

barmac replied:
Since `outgoing` is a `Map`, `outgoing.keys()` returns an `Array`, and `[].map` creates a new `Array`, we don't need to do either `[ ...outgoing.keys() ]` or `Array.from(outgoing.keys())`. It's completely safe and IMO easier to read if we just do `outgoing.keys().map(key => graph.get(key))`.

barmac replied:
Oh OK. Thanks for clarification and sorry for BS

marstamm replied:
No worries, I assumed it creates an array at first as well. It can be confusing, especially since `Object.keys` returns an array 😄
